### PR TITLE
Added Pigeonhole sort

### PR DESCRIPTION
**Pull request by VishnuPothan**
# Welcome to Dart community

### **Describe your change:**

* [x] Add an algorithm?
* [ ] Fix a bug or typo in an existing algorithm?
* [ ] Documentation change?

### **Checklist:**
* [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Dart/blob/master/CONTRIBUTING.md).
* [x] This pull request is all my own work -- I have not plagiarized.
* [x] I know that pull requests will not be merged if they fail the automated tests.
* [x] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [x] All new Dart files are placed inside an existing directory.
* [ ] All new algorithms have a URL in its comments that points to Wikipedia or other similar explanation.
* [ ] If this pull request resolves one or more open issues then the commit message contains `Fixes: #{$ISSUE_NO}`.

### CODE REVIEW

File: sort/pigeonhole_sort.dart [hunk 1]
```diff
@@ -0,0 +1,42 @@
+void pigeonholeSort(List arr) {
```
Comment (Parowicz) [inline sort/pigeonhole_sort.dart:1]: Would you mind to add some tests?
```suggestion
import 'package:test/test.dart';

void pigeonholeSort(List arr) {
```

File: sort/pigeonhole_sort.dart [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+}
+
+void main() {
+  List list = [87, 48, 5, 7, 135, 85];
+  print('Before sorting:');
+  print(list);
+  print('--------------------------------------');
+  print('After sorting:');
+  pigeonholeSort(list);
+  print(list);
```
Comment (Parowicz) [inline sort/pigeonhole_sort.dart:41]: Also seems like `pigeonholeSort` have some problems when empty list is provided.

```suggestion
  test("Sort empty list returns empty list", () {
    List list = [];
    pigeonholeSort(list);
    expect(list, isEmpty);
  });

  test("Already sorted list remain sorted", () {
    List list = [1, 2, 3, 4, 5];
    pigeonholeSort(list);
    expect(list, equals([1, 2, 3, 4, 5]));
  });

  test("Sort", () {
    List list = [87, 48, 5, 7, 135, 85];
    pigeonholeSort(list);
    expect(list, equals([5, 7, 48, 85, 87, 135]));
  });

  test("Sorted list size doesnt change", () {
    List list = [1, 1, 4, 1, -12, -12, 77];
    pigeonholeSort(list);
    expect(list.length, equals(7));
    expect(list, [-12, -12, 1, 1, 1, 4, 77]);
  });
```

**Comment by VishnuPothan** [comment]
@Parowicz Fixed the warning on empty list.

File: sort/pigeonhole_sort.dart (hunk 2 above)
Comment (Parowicz) [reply sort/pigeonhole_sort.dart:48]: Do you wish to add those test cases?

**Comment by VishnuPothan** [comment]
@Parowicz added the test cases.

**Review by Parowicz — APPROVED** [review]
LGTM!

**Pull request merged**